encoding/codecv3/codecv3.go

```diff
@@ -0,0 +1,243 @@
+package codecv3
+
+import (
+	"encoding/binary"
+	"encoding/hex"
+	"errors"
+	"fmt"
+
+	"github.com/scroll-tech/go-ethereum/common"
+	"github.com/scroll-tech/go-ethereum/crypto"
+	"github.com/scroll-tech/go-ethereum/crypto/kzg4844"
+
+	"github.com/scroll-tech/da-codec/encoding"
+	"github.com/scroll-tech/da-codec/encoding/codecv1"
+	"github.com/scroll-tech/da-codec/encoding/codecv2"
+)
+
+// DABlock represents a Data Availability Block.
+type DABlock = codecv2.DABlock
+
+// DAChunk groups consecutive DABlocks with their transactions.
+type DAChunk = codecv2.DAChunk
+
+// DABatch contains metadata about a batch of DAChunks.
+type DABatch struct {
+	// header
+	Version                uint8
+	BatchIndex             uint64
+	L1MessagePopped        uint64
+	TotalL1MessagePopped   uint64
+	DataHash               common.Hash
+	BlobVersionedHash      common.Hash
+	ParentBatchHash        common.Hash
+	LastBlockTimestamp     uint64
+	BlobDataProof          [64]byte
+	SkippedL1MessageBitmap []byte
+
+	// blob payload
+	blob *kzg4844.Blob
+	z    *kzg4844.Point
+}
+
+// NewDABlock creates a new DABlock from the given encoding.Block and the total number of L1 messages popped before.
+func NewDABlock(block *encoding.Block, totalL1MessagePoppedBefore uint64) (*DABlock, error) {
+	return codecv2.NewDABlock(block, totalL1MessagePoppedBefore)
+}
+
+// NewDAChunk creates a new DAChunk from the given encoding.Chunk and the total number of L1 messages popped before.
+func NewDAChunk(chunk *encoding.Chunk, totalL1MessagePoppedBefore uint64) (*DAChunk, error) {
+	return codecv2.NewDAChunk(chunk, totalL1MessagePoppedBefore)
+}
+
+// NewDABatch creates a DABatch from the provided encoding.Batch.
+func NewDABatch(batch *encoding.Batch) (*DABatch, error) {
+	// this encoding can only support a fixed number of chunks per batch
+	if len(batch.Chunks) > codecv2.MaxNumChunks {
+		return nil, fmt.Errorf("too many chunks in batch")
+	}
+
+	if len(batch.Chunks) == 0 {
+		return nil, fmt.Errorf("too few chunks in batch")
+	}
+
+	if len(batch.Chunks[len(batch.Chunks)-1].Blocks) == 0 {
+		return nil, fmt.Errorf("too few blocks in last chunk of the batch")
+	}
+
+	// batch data hash
+	dataHash, err := codecv1.ComputeBatchDataHash(batch.Chunks, batch.TotalL1MessagePoppedBefore)
```

Comment:
I'm quite confused that codev3 calls codev1&codev2 methods. maybe we should do high level abstract if possible.

Reply:
refactored the misuse in [da72524](https://github.com/scroll-tech/da-codec/pull/11/commits/da72524ebc6e38689dd8d0c373082340ee41d3b6).